src/mux_client_impl/child.rs

```diff
@@ -0,0 +1,112 @@
+use super::{ChildStderr, ChildStdin, ChildStdout, Error};
+
+use core::mem::replace;
+
+use std::io;
+use std::os::unix::process::ExitStatusExt;
+use std::process::ExitStatus;
+
+use openssh_mux_client::connection::{EstablishedSession, SessionStatus};
+
+#[derive(Debug)]
+pub struct RemoteChild {
+    state: RemoteChildState,
+    child_stdin: Option<ChildStdin>,
+    child_stdout: Option<ChildStdout>,
+    child_stderr: Option<ChildStderr>,
+}
+
+impl RemoteChild {
+    pub(crate) fn new(
+        established_session: EstablishedSession,
+        child_stdin: Option<ChildStdin>,
+        child_stdout: Option<ChildStdout>,
+        child_stderr: Option<ChildStderr>,
+    ) -> Self {
+        Self {
+            state: RemoteChildState::Running(established_session),
+            child_stdin,
+            child_stdout,
+            child_stderr,
+        }
+    }
+
+    pub async fn disconnect(self) -> io::Result<()> {
+        use RemoteChildState::*;
+
+        match self.state {
+            Intermediate => unreachable!(),
+            Exited(_exit_status) => Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "invalid argument: can't kill an exited process",
+            )),
+            Running(_established_session) => Ok(()),
+        }
+    }
+
+    pub async fn wait(&mut self) -> Result<ExitStatus, Error> {
+        use RemoteChildState::*;
+
+        let exit_value = match self.state.take() {
+            Intermediate => unreachable!(),
```

Comment:
I don't think that can happen.

Rusts' `async` feature is lazy by design.
Until the point it is `.await`ed, nothing happens.

The returned `Future` just stores the arguments passed to the function.

Reply:
It's still possible because a future can be cancelled _at any time_. So, for example:
```rust
async fn oops(child: &mut RemoteChild) {
  let wait = child.wait();
  let cx = /* make my own Context */;
  wait.poll(cx);
  drop(wait);
  child.wait().await;
}
```
Or, again, with `select!`, this can totally happen, since `select!` also can cancel a future _at any time_, including after having polled it once.

Reply:
Which is also why I believe the state should be called `AwaitingExit`, since that's what it actually represents.

Reply:
Thank you for reminding me that.

I forgot that Future can be cancelled.

Reply:
Variant renamed.